src/trackedit/internal/au3/au3interaction.cpp

```diff
@@ -1394,30 +1394,21 @@ void Au3Interaction::undo()
 
     auto trackeditProject = globalContext()->currentProject()->trackeditProject();
 
-    // Find the index of the currently selected track in the old list
-    auto trackIdList = trackeditProject->trackIdList();
-    auto selectedTracks = selectionController()->selectedTracks();
-    int selectedIndex = -1;
-    if (!selectedTracks.empty()) {
-        selectedIndex = std::distance(trackIdList.begin(),
-                                      std::find(trackIdList.begin(),
-                                                trackIdList.end(),
-                                                selectedTracks.at(0)));
-    }
     projectHistory()->undo();
 
     // Undo removes all tracks from current state and
     // inserts tracks from the previous state so we need
     // to reload whole model
     trackeditProject->reload();
 
-    // Update selected track id
-    auto newTrackIdList = trackeditProject->trackIdList();
-    if (selectedIndex >= 0 && selectedIndex < static_cast<int>(newTrackIdList.size())) {
-        selectionController()->setSelectedTracks(newTrackIdList);
-    } else {
-        selectionController()->resetSelectedTracks();
+    // Select tracks that were previously selected
+    auto& project = projectRef();
+    auto& tracks = Au3TrackList::Get(project);
+    TrackIdList selectedTracks;
+    for (const auto & selectedTrack : tracks.Selected()) {
+        selectedTracks.push_back(DomConverter::trackId(selectedTrack->GetId()));
     }
+    selectionController()->setSelectedTracks(selectedTracks);
```

Comment:
@grliszas14 here you sync selectionController state with its internal data model (tracks.Selected), `Au3SelectionController::setSelectedTracks` is doing exactly that it sets each track to its current `selected` state, which is probably not what we want to do.

 I'm not 100% sure, but i probably would make more sense if we update selectionController state when tracklist sends SELECTION_CHANGE

Reply:
yes, it sets each track to its current selected state but it's additionally updating its own state (m_selectedTracks), yes, it would make more sense to respond to SELECTION_CHANGE if we got SELECTION_CHANGE.
undo/redo replaces whole model (which means getting multiple TrackListEvent::DELETION/ADDITION) - the problem I see here is that every DELETION event has trackId=-1 -> so I can't really respond to that because I don't know which track to delete for example

Reply:
Right, that's unfortunate. So, how about hook selectionController to TrackListEvent::DELETION/ADDITION to update its state?

As I understand there are other situations where tracklist gets changed besides undo/redo, and it will also require us to update selectionController state, right?